runtime/sema/check_composite_declaration.go

```diff
@@ -1410,6 +1540,37 @@ func (checker *Checker) checkTypeRequirement(
 
 	requiredInterfaceType := requiredCompositeType.InterfaceType()
 
+	// while attachments cannot be declared as interfaces, an attachment type requirement essentially functions
+	// as an interface, so we must enforce that the concrete attachment's base type is a compatible with the requirement's.
+	// Specifically, attachment base types are contravariant; if the contract interface requires a struct attachment with a base type
+	// of `S`, the concrete contract can fulfill this requirement by implementing an attachment with a base type of `AnyStruct`:
+	// if the attachment is valid on any structure, then clearly it is a valid attachment for `S`. See the example below:
+	//
+	// resource interface RI { /* ... */ }
+	// resource R: RI { /* ... */ }
+	// contract interface CI {
+	//    attachment A for R { /* ... */ }
+	// }
+	// contract C: CI {
+	//    attachment A for RI { /* ... */ }
+	// }
+	//
+	// In this example, as long as `A` in `C` contains the expected member declarations as defined in `CI`, this is a valid
+	// implementation of the type requirement, as an `A` that can accept any `RI` as a base can clearly function for an `R` as well.
+	// It may also be helpful to conceptualize an attachment as a sort of implicit function that takes a `base` argument and returns a composite value.
+	if requiredCompositeType.Kind == common.CompositeKindAttachment && declaredCompositeType.Kind == common.CompositeKindAttachment {
+		if !IsSubType(requiredCompositeType.baseType, declaredCompositeType.baseType) {
+			checker.report(
+				&ConformanceError{
```

Comment:
It might be good to extend `ConformanceError` and provide the reason for the non-conformance, i.e. the base type mismatch. See the other similar fields in `ConformanceError`.

A generic error that the concrete type does not conform to the type requirement is hard to debug, especially if the reason is that the base types of some attachment don't match.

Reply:
I agree, but IMO this is a separate PR; adding reasons to conformance errors is more widespread a change than just attachments.

Reply:
Yes, in a separate PR is preferred 👍